### PR TITLE
ResolveTargetEntityListener - not loadClassMetadata on each request fix #230

### LINKED ISSUE #230: ResolveTargetEntityListener and Doctrine\ORM\Event::loadClassMetadata each request
I use RTEL and there are huge performance problems, because Doctrine\ORM\Event::loadClassMetadata is called on each request. If I look into cached files, there are correct targetEntity and so imho that calling is useless...

events

### PR DESCRIPTION
**Pull request by jirinapravnik**
This PR should fix #230 . There was some kind of custom resolving metadata for target entity in ClassMetadataFactory, but Doctrine itself imho has better and faster resolving via onClassMetadataNotFound event. I tried it on local project and it is realy faster. Maybe there should be better tests for testing if is fired really this event during resolving, but I am not able to create it.

### CODE REVIEW

File: src/Kdyby/Doctrine/Tools/ResolveTargetEntityListener.php [hunk 1]
```diff
@@ -29,7 +29,10 @@ class ResolveTargetEntityListener extends Doctrine\ORM\Tools\ResolveTargetEntity
 	 */
 	public function getSubscribedEvents()
 	{
-		return array(Events::loadClassMetadata);
+		return array(
+			Events::loadClassMetadata,
+			Events::onClassMetadataNotFound
+		);
```
Comment (enumag) [inline src/Kdyby/Doctrine/Tools/ResolveTargetEntityListener.php]: This method can be removed - it's the same as in the parent class.

File: src/Kdyby/Doctrine/Events.php [hunk 2]
```diff
@@ -116,6 +116,14 @@
 	const loadClassMetadata = 'Doctrine\\ORM\\Event::loadClassMetadata';
 
 	/**
+	 *  The onClassMetadataNotFound event occurs whenever loading metadata for a class
```
Comment (enumag) [inline src/Kdyby/Doctrine/Events.php]: Double space at the beginning.

File: src/Kdyby/Doctrine/Mapping/ClassMetadataFactory.php [hunk 3]
```diff
@@ -70,10 +70,7 @@ protected function loadMetadata($name)
 			throw new Kdyby\Doctrine\MissingClassException("Metadata of class $name was not found, because the class is missing or cannot be autoloaded.");
 		}
 
-		$result = parent::loadMetadata($name);
-		if ($name !== $origName) {
-			$this->setMetadataFor($origName, $this->getMetadataFor($name));
-		}
+		$result = parent::loadMetadata($origName);
```
Comment (enumag) [inline src/Kdyby/Doctrine/Mapping/ClassMetadataFactory.php]: The variable is no longer needed, just use `return parent::loadMetadata($origName);`.

File: src/Kdyby/Doctrine/Tools/ResolveTargetEntityListener.php [hunk 4]
```diff
@@ -29,7 +29,10 @@ class ResolveTargetEntityListener extends Doctrine\ORM\Tools\ResolveTargetEntity
 	 */
 	public function getSubscribedEvents()
 	{
-		return array(Events::loadClassMetadata);
+		return array(
+			Events::loadClassMetadata,
+			Events::onClassMetadataNotFound
```
Comment (enumag) [inline src/Kdyby/Doctrine/Tools/ResolveTargetEntityListener.php]: I would add a comma to the end of this line.

File: tests/KdybyTests/Doctrine/TargetEntityMapping.phpt [hunk 5]
```diff
@@ -56,6 +57,20 @@ class TargetEntityMapping extends KdybyTests\Doctrine\ORMTestCase
 		Assert::same('KdybyTests\Doctrine\CmsAddress', $meta->name);
 	}
 
+	public function testListenersExists(){
```
Comment (enumag) [inline tests/KdybyTests/Doctrine/TargetEntityMapping.phpt]: The `{` should be on new line.

File: tests/KdybyTests/Doctrine/TargetEntityMapping.phpt [hunk 6]
```diff
@@ -56,6 +57,20 @@ class TargetEntityMapping extends KdybyTests\Doctrine\ORMTestCase
...
+		$evm = $this->serviceLocator->getByType('Kdyby\Events\EventManager');
+		$loadClassMetadata = $evm->getListeners(Events::loadClassMetadata);
+		$onClassMetadataNotFound = $evm->getListeners(Events::onClassMetadataNotFound);
+
+		$filterRTEL = function($items){
+			return array_filter($items, function($item){
```
Comment (enumag) [inline tests/KdybyTests/Doctrine/TargetEntityMapping.phpt]: Should be `function ($var) {` (missing spaces).

**Comment by enumag** [comment]
The test doesn't feel right - it tests that the listener is registered correctly but not that it works as it should.

**Comment by jirinapravnik** [comment]
@enumag Thanks for review, it should be fixed. Also I tried better test. I also left test for correctly registered listeners, imho it should be there. But in fact listeneres are also indirectly tested in new test.  Should I remove testListenersExists? What do you mean?

**Comment by enumag** [comment]
@jirinapravnik The problem is (in my opinion) that the tests don't test the buggy use case that caused you to send this PR. The test I'd like to see should ensure that for multiple calls of getClassMetadata the loadClassMetadata would only be triggered once. And we would have to make sure that the test would really fail with previous version of Kdyby/Doctrine.

**Comment by jirinapravnik** [comment]
@enumag 
Previous Kdyby\Doctrine. I have
IAdministrator
BaseAdministrator implements IAdministrator //MappedSuperclass
StandardAdministrator extends BaseAdministrator //targetEntity

It calls loadClassMetadata for BaseAdministrator and for StandardAdministrator. It calls this method: https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php#L97 (it also remap associations etc, it is useless, it is in cache yet)

But if onClassmetadataNotFound event it calls faster method: https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php#L77 . And it calls for interface.

Problem is not in two calls of loadClassMetadata. The problem is imho here: https://github.com/Kdyby/Doctrine/blob/master/src/Kdyby/Doctrine/Mapping/ClassMetadataFactory.php#L62 . It changes interface to entity, then imho it calls again and again for class implementing interface and not just once for interface.

The test fails with previous version, because it doesn't fire onClassMetadataNotFound event... And it is what is necessary for this faster method.

**Comment by enumag** [comment]
Imho the loadClassMetadata event should NOT be called twice. Only the first time when data are not in cache.

What triggers the loadClassMetadata event in every request? I'm guessing its some call to getClassMetadata, but with what argument?

**Comment by jirinapravnik** [comment]
@enumag Prejdu do cestiny, snad nevadi, anglictina se neumim poradne vyjadrit:)

Problem je nejspise v tomhle. Aspon to tak chapu. Ze soucasna implementace ClassMetadataFactory v Kdyby vezme, pokud je to interface, vezme si tu targetEntity a haze loadClassMetadata pro ni. Kdezto korektni chovani (zkousel jsem to v symfony, jak to funguje), tak tam se na interface haze event onClassmetadataNotFound, ktery se zavola pro kazdy interface jednou a pak je tam nekde ulozeny a uz se nevyhazuje.

Ve stavajici verzi se tedy nevyhazuje vickrat loadClassMetadata, ale vyhazuje se vzdy pro kazdou tridu implementujici ten interface. Coz je logicky nesmysl, proc vyhazovat loadClassMetadata pro BaseEntitu, z ktere pak odvozuji tu dalsi entitu apod...

Jednoduse 
1 interface, 1base entita implementujici interface, 1 entita extends baseentita
nyni - 2x loadClassMetadata
nove a i podle symfony (resp. ciste doctrine) korektne - 1x onClassNotFoundMetadata, ktery je i vykonove rychlejsi

**Comment by enumag** [comment]
V tom případě jak vzniklo to že se ti loadClassMetadata pustila při každém requestu?

**Comment by jirinapravnik** [comment]
Při každém requestu se má spustit onClassNotFoundMetadata (podle mě by bylo lepší kdyby tomu tak nebylo, ale je to asi očekávané, v symfony se tomu tak taky děje). No a vzhledem k tomu jak píšu výše, když to nevystřelilo tohle, tak to furt vystřelovalo loadClassMetadata.

**Comment by enumag** [comment]
Jestli dobře chápu zdrojáky tak to co dělalo Kdyby mělo za cíl právě to aby se ten event onClassNotFoundMetadata nevyvolával pořád - akorát to z nějakého důvodu nefunguje.

**Comment by jirinapravnik** [comment]
Jo to by bylo fajn, kdyby to fungovalo... Pokládám si však otázku, zda je to vůbec nějak řešitelné, protože v Doctrine je ten notFound event jenom v kontextu ResolveTargetEntityListener, jinde používám není...

Mě se ještě podařilo, že jsem prohodil tu podmínku na nové a staré jmeno s tím $result = parent::loadClassMetadata() a to fungovalo lépe, nicméne furt to vystřelovalo pro MappedSuperclass loadClassMetadata event, protože tady mají listener tyhle třídy: https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L240

**Comment by enumag** [comment]
Dokázal bys napsat failující test bez použití toho eventu onClassNotFoundMetadata?

**Comment by jirinapravnik** [comment]
Ať dělám co delám, tak nejsem... Ale ja nejsem moc zkušený tester, takže je možný, že ostatní ano.

**Comment by enumag** [comment]
Myslel jsem si to. Sám to asi také nezvládnu protože nemám k dispozici aplikaci kde bych tu chybu viděl takže bych to jen těžko dával dohromady. Co kdybychom to vyřešili společně přes teamviewer?

**Comment by jirinapravnik** [comment]
To by asi bylo celkem slozite. Vypreparoval jsem nejaky casti a pouzil to na sandbox a je to tam viditelne.
https://github.com/jirinapravnik/doctrine-rtel
composer install nainstaluje ten muj update - tam vidis, ze je zavolan ten even notFound, kdyz zmenis na dev-master - dostanes master kdyby samo

databaze k tomu by nemeal byt treba volam jen getClassMetadata

ale to ti jiste nemusim rikat

mrkni pak v kdyby/events panelu - kdyz to prepnes na master, tak tam uvidis, ze celkem nesmyslne to vola loadClassMetadata na Standard i na BaseTag

**Comment by enumag** [comment]
Díky, mrknu na to zítra.

**Comment by enumag** [comment]
Hmm... proč se ptáš na metadata interfacu? Pokud správně chápu ClassMetadataFactory tak se zabývá pouze třídami, nikoli rozhraními. Zatím vidím jen bug v tom že ClassMetadataFactory z Kdyby nehodí výjimku když se zeptáš na interface.

**Comment by matej21** [comment]
co takhle to getTargetEntityClassName presunout z loadMetadata do getMetadataFor? jen se teda kvuli performance bude muset zduplikovat ta loadedMetadata cache

**Comment by fprochazka** [comment]
Od kdy tam je ten `onClassNotFoundMetadata` ? Fakt je to rychlejší? Není blbý že se to volá na každej request? To by se mělo ideálně zacachovat pro ten interface ty metadata a pak už nevolat, ne?

**Comment by enumag** [comment]
Ten event je novinka v 2.5.0, nevím přesně proč. Taky je mi divný že by se měl volat pořád...

**Comment by fprochazka** [comment]
@enumag event je určitě elegantnější řešení než ten hack co tam byl předtím (který jsem si celkem jistej že jsme kopírovali ze symfony). Ale aby se volal na každej rq je blbost.

**Comment by jirinapravnik** [comment]
@fprochazka Nemůžu říct zatím z produkce jak je to lepší, chci to nasadit raději po Novém roce. Nicméně když jsem to měl v produkci, tak stačilo u webu kam chodí 10tisíc lidi za den (ci-li nic extra bych rekl) a kazdou chvili sesypany web, cas jednonasobne delsi zpracovani nez kdyz jsem vyradil interaces... Zkrátka to furt načítalo znovu metadata, remapovalo asociace atd. i když to bylo nacachovaný. Tady ten nový event, to v podstatě vezme a přiřadí k interfacu načtený metadata te targetEntity. Pokud tomu dovře rozumím, tohle je to zpracování eventu (https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php#L77) a v podstatě dělá to samé, co se tam snažíme hackovat, či-li tady je to časově navíc jen vystřelení a chycení eventu oproti tomu, když děláme to přiřazení hackem.

Zkoušel jsem si odpoledne hrát s tím, že jsem tu funkcionalitu přesunul do getMetadataFor a vypadalo to funkčně a nevystřeloval se notFound event. Ale je otazka zda to řešit na úrovni Kdyby, zda by nebylo lepší to pořešit nějak lépe na straně Doctrine, ale tam si netroufnu...

**Comment by jirinapravnik** [comment]
@fprochazka bylo by mozne to mergnout, nebo planujes nejake jine reseni? od te doby, co jsem to nasadil zadne problemy vykonostni a bezi to i na dalsich webech uz. Ta aktualni verze Kdyby\Doctrine dela tyhle  Interface Entity nepouzitelne na nejakem projektu rozumne velikosti...

**Comment by enumag** [comment]
@jirinapravnik Můžeš to zkusit s novou verzí Kdyby/DoctrineCache? Teoreticky to mohl způsobovat tenhle bug: https://github.com/Kdyby/DoctrineCache/commit/c15b87fdfa8a3becf29965d2670c4fd52134b515.

**Comment by jirinapravnik** [comment]
@enumag žádná změna...

**Comment by fprochazka** [comment]
Thank you for solving this! And sorry for taking too long.

**Pull request closed without merging**